### ISSUE TITLE
unfriendly object deconstruction format after 2.3.0

### ISSUE THREAD

**Issue by JounQin** [prettier/prettier#10866]
**Prettier 2.3.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEwA6VOYCCAMoQKoAihA+hQJIDKhAQsQKLUDChnAEmwBo8BEuSq1GLdtVIAtAJpcA8gFkACnXZD8mNgA0AKmwBKAOVFq12gL6YAvJgAOAJwhg4aNADoEANxACIBCOMACW6MigAIbOrgDuajEIaMggUQA2cVEAnimBAEbOUWAA1nAwDI7FoVAA5sgwzgCucIFwALb5cAAm3T3EUXVNUbVwAGIQzu1RMGF1qVFNMBABIAAWMO3pAOprofBoVe4MyfuhvvvZqWCeqzVocM4wakW108gAZhkPgQBWaAAPZhFUrlBhRdpwYg1OCfb6tED-AEMGq1dJwACKTQg8Dh6R+ICqzgezlSMGyjg8YGcoRCqxcNRg21C3Rga2QAA4AAyBFwQB7bIqOVIuDyPXywwIAR2x8BewRSKCiaAAtFA4D0eqtnHAZaEdS8Ru8kF98QiHu1Qg1mubUeisTjYSb4YEYFF8szWeykAAmV1FULpVGcCDtY0gDwAVlWTQeBnditNBN8LToUD6sAY1NpMEI6YY5PReIe1msQA)

**Input:**

```jsx
const {
  ALAUDA_DISABLE_CACHE,
  ALAUDA_DISABLE_LAZY_COMPILE,
  EXTERNALAPP,
} = process.env
```

**Output:**

```jsx
const { ALAUDA_DISABLE_CACHE, ALAUDA_DISABLE_LAZY_COMPILE, EXTERNALAPP } =
  process.env;
```

**Expected behavior:**

Don't change.

**Comment by JounQin** [comment]
related #10850, #10857

**Comment by thorn0** [comment]
The destructuring fits on the line and is a simple comma-separated list of identifiers. Why break it? Imagine the identifiers are `x, y, z` instead of `ALAUDA_DISABLE_CACHE, ALAUDA_DISABLE_LAZY_COMPILE, EXTERNALAPP`. In your opinion, would it still make sense to break the destructuring only because the right-hand side of the assignment doesn't fit on the line?

**Comment by JounQin** [comment]
I think the key point is whether the value is printed prefixed with `} =`.

For short statement, I prefer `const {a, b} = process.env`, for long statement, maybe the following is also fine to me.

```js
const {
  ALAUDA_DISABLE_CACHE, ALAUDA_DISABLE_LAZY_COMPILE, EXTERNALAPP
  } = process.env;
```

**Comment by thorn0** [comment]
> the key point is whether the value is printed prefixed with `} =`

Interesting. Why do you think so?

**Comment by JounQin** [comment]
It obviously indicates that it is an object reconstruction statement IMO. I only need read a few charters to guess/confirm that, but in current implementation I have to look through the whole line.

**Comment by PupoSDC** [comment]
I agree, the new syntax really shifts focus from the most important information.

```ts
  const {
    isBuffering,
    currentTime,
    duration,
    setTrack,
    play,
    pause,
  } = useAudioPlayer(props);
```

became this

```ts
  const { isBuffering, currentTime, duration, setTrack, play, pause } =
    useAudioPlayer(props);
```

Which reads extremely award, since unless you check the line above to understand what is happening something like:

```ts
  const {
    isBuffering, currentTime, duration, setTrack, play, pause
  } = useAudioPlayer(props);
```

could be acceptable, but in terms of readability and git diffing the original version is by far the clearest.

**Comment by PupoSDC** [comment]
Unfortunately still an issue with `2.3.2`

https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEwA6VOYCCAMoQKoAihA+hQJIDKhAQsQKLUDChnAEmwBo8BEuSq1GLdtVIAtAJpcA8gFkACnXZD8mNgA0AKmwBKAOVFq12gL6YAvJgAOAJwhg4aNADoEANzwgAiAQjjAAlujIoACGzq4A7mqxCGjIINEANvHRAJ6pQQBGztFgANZwMAyOJWFQAObIMM4ArnBBcAC2BXAAJj29xNH1zdF1cABiEM4d0TDh9WnRzTAQgSAAFjAdGQDq62HwaNXuDCkHYb4HOWlgnmu1aHDOMGrFdTPIAGaZj0EAVmgAB7MYplCoMaIdODEWpwL4-NogAGAhi1OoZOAARWaEHg8IyvxA1Wcj2caRgOUcHjAzjCoTWLlqMB2YR6MHWyAAHAAGIIuCCPHbFRxpFweJ6+OFBACOOPgrxCqRQ0TQAFooHBer01s44LKwrrXqMPkhvgTEY8OmFGi0LWiMdjcXDTQigjBogUWWyOUgAExu4phDJozgQDomkAeACsa2ajwMHqVZsJvladCg-VgDBpdJghAzDApGPxj2s1iAA

**Comment by u3u** [comment]
Any news? This one is really kind of ugly 😭

**Comment by rodriguezmanu** [comment]
+1

**Comment by u3u** [comment]
Now I solved it for now by manually adding the `//` comment
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAExMHpcA6UmmAlmgEICuAZjXAE6lQDmANJmFQwwjACqkAtnA4ATbgEMYpaBzRwBDSWADWHAA4AbSQE9NkqgswBfTAF4iJI3ACCVMbIAKO3YwAUGhhA1oAlADcRCBsID4y6MigkjwQAO5OMQhoyCCSWnF6KaEARspqigDKkiIAMsxwyDTpCrn5qkUaKswsyDAMVKIgcEI5cGJi-aWSrFSSLHAAYhAMQtIyrKmGMBAhIAAWMEJaAOrrpPBoTWBwhckHpABuB7qpYGjZIMwKDDBOyixzVTVdAFZoAA9Ci0tHAAIpUCDwb5aWogJoMF6pGC6DRwNBgJgaGBrLzMGA7UhiGDrZAADgADKEvBAFDtlBpUl50YxLpVQgBHSHwd4+FIoSRoAC0UDg-X6a14XNIvHe4y+SGqsK6CiEpDaHRVIPB3Mqip+oRgkhyhOJpKQACZDcpSFoWgBhCBCBXdNAAVjWNn4xv5Srhl06AEkoINYIVMaRsbYQ4UUaCYXD-gCKPUiiU4OVRQm4CYTEA)

**Input:**

```ts
const { //
  isBuffering, currentTime, duration, setTrack, play, pause } =
  useAudioPlayer(props);
```

**Output:**

```ts
const {
  //
  isBuffering,
  currentTime,
  duration,
  setTrack,
  play,
  pause,
} = useAudioPlayer(props);
```

**Comment by thorn0** [comment]
Any specific proposals on how to generalize your suggestions and improve the current heuristics? It might be something like this (just examples):

"If some of the identifiers consist of more than 3 words, always break the destucturing even if the whole statement fits on one line."

or

"If there is more than 5 elements in the destructuring, prefer to break the left side of the assignment, but keep the statement on one line if it still fits."

etc

Achieving a satisfying result in this case is a matter of playing with heuristics and finding a good enough one.

**Comment by PupoSDC** [comment]
I would suggest following the same rules as we currently do for object definition [as explaiend here](https://prettier.io/docs/en/rationale.html#multi-line-objects).

```ts
const { potato, 
 kiwi
} = fruits;
```

is formatted into 

```ts
const { potato, kiwi } = fruits;
```

while 

```ts
const { 
 potato,  kiwi
} = fruits;
```

is formatted to 

```ts
const { 
 potato,  
 kiwi
} = fruits;
```

**Comment by thorn0** [comment]
@PupoSDC That rules is only a workaround. See https://prettier.io/docs/en/rationale.html#%EF%B8%8F-a-note-on-formatting-reversibility
Why are you giving up without even trying to find a solution?

**Comment by PupoSDC** [comment]
giving up? 😕

**Comment by thorn0** [comment]
Requests to keep the original formatting are out of scope for Prettier.

**Comment by PupoSDC** [comment]
To make my suggestion a bit clear... What Im proposing is that object destructuring is treated the same way as the assignment. When a solution is found for the second, the same solution should be used in the first. Im not suggesting to "keep the original format" and I'm most definitely not "giving up". 🙄 

My personal opinion is that any statement that goes over the line limit should be expanded as a multiline statement on the "object" side of the equation

```ts
// orignal input, breaking line width
const { alpha, bravo, charlie, delta, echo, foxtrot, golf, hotel, india } = alphabet
const b = { alpha, bravo, charlie, delta, echo, foxtrot, golf, hotel, india } 

// should be resolved to
const { 
  alpha, 
  bravo, 
  charlie, 
  delta, 
  echo, 
  foxtrot, 
  golf, 
  hotel, 
  india 
} = alphabet
const b = { 
  alpha, 
  bravo, 
  charlie, 
  delta, 
  echo, 
  foxtrot, 
  golf, 
  hotel, 
  india 
}
```

**Comment by thorn0** [comment]
If you implement your suggestion as a PR and use this project https://github.com/prettier/prettier-regression-testing to check the results, my guess is that you'll see a lot of unbroken right sides of assignments that look much better when the left side is on one line. This logic has to be more discreet.

**Comment by mmeqbel-wiley** [comment]
i couldn't integrate prettier into our project because of this issue
